### PR TITLE
Add Option: ignoreHeadersWhenXDR

### PR DESCRIPTION
**Pull request by simon-wenmouth**
Add Option: ignoreHeadersWhenXDR

Allow caller to decide if it is an error to pass HTTP headers to CORS requests in IE<10.
    
This is useful in scenarios when the library is being used to make XDR calls and handle the difference between legacy and modern browsers (i.e. transparently choose to use either XMLHttpRequest or XDomainRequest as appropriate on behalf of the library user).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -200,7 +200,9 @@ function _createXHR(options) {
             }
         }
     } else if (options.headers && !isEmpty(options.headers)) {
-        throw new Error("Headers cannot be set on an XDomainRequest object")
+        if (('ignoreHeadersWhenXDR' in options) == false || !!!options.ignoreHeadersWhenXDR) {
```
Comment (naugtur) [inline index.js]: `!!!` seems loud. is it any different from using just one? 
Also, `('ignoreHeadersWhenXDR' in options) == false` doesn't seem different from just `!options.ignoreHeadersWhenXDR`

**Comment by naugtur** [comment]
I don't see why instead of passing
```
headers: {...},
ignoreHeadersWhenXDR: shouldIgnore
```

you couldn't pass

```
headers: (shouldIgnore)? null, {...}
```

This issue was raised a long time ago and we decided to let client pass headers conditionally instead of adding a non-standard option that is hard to explain.

BTW. Take a look at our goals for next major version - we're planning to create a legacy version for old IEs, so a feature like this would make more sense there. I'd be happy if you wanted to help out. The most urgent thing to do is setting up cross-browser testing. 

Also, my condolences on working with IE<10 ;)

**Comment by simon-wenmouth** [comment]
In my use cases the HTTP headers are optional for IE<10 (e.g. content-type or correlation-id) and as such it is desirable for my applications to always pass headers and still succeed if the request falls back to XDR on IE<10 and simply not transmit HTTP headers.  My original approach was to remove the block with `throw Error` but I did not want to introduce a breaking change for existing users.

A non-breaking alternative would be an option that allows the caller to choose between:

(a) throwing an error (default setting; current behavior),

(b) to issue a callback so callers have a single error-path:

```
var response = failureResponse,
      err = new Error("Headers cannot be set on an XDomainRequest object");
return callback(err, response, response.body);
```

(c) ignore the headers (the behavior in this PR)

For cross-browser testing, we're you thinking of something like [sauce labs](https://saucelabs.com/open-source)?  I'd be happy to help out.

**Comment by naugtur** [comment]
>  to always pass headers and still succeed if the request falls back to XDR on IE<10 and simply not transmit HTTP headers

But I don't understand why `headers: window.XDomainRequest?null:{...}` doesn't solve your case. You can just make a decision to skip headers outside `xhr`.

As for tests - our biggest issue is that for functional tests we need a backend. I tried making requests to mock servers online, but the test environment didn't have outbound connection capability. Now we're using a server on localhost, so the env would have to run node and the right browser. This is getting complicated really fast. 
I'm looking at leveraging saucelabs, but I don't yet know how to let the tests connect to a mock server of any kind.

**Comment by simon-wenmouth** [comment]
You're right, I can pass null (as you suggest) to avoid the error.  It's just unfortunate that XDomainRequest has snuck out from behind `xhr`.

As a side note, when the code falls back to XDR the code actually modifies the input header object.

```
var headers = xhr.headers = options.headers || {}
// ... snip ...
if ("json" in options) {
  isJson = true
  headers["accept"] || headers["Accept"] || (headers["Accept"] = "application/json") // << here
  if (method !== "GET" && method !== "HEAD") {
    headers["content-type"] || headers["Content-Type"] || (headers["Content-Type"] = "application/json") // << here
    body = JSON.stringify(options.json)
  }
}
```

As such, when `options.headers = {}` and `options.json = true` then `(options.headers && !isEmpty(options.headers))` will always evaluate to true and throw the error (which is how i stumbled upon this edge case).

I know Travis-CI allows you to spin up a [web-server](https://docs.travis-ci.com/user/gui-and-headless-browsers/#Starting-a-Web-Server) (e.g. node.js) before tests run and has [sauce labs](https://docs.travis-ci.com/user/gui-and-headless-browsers/#Using-Sauce-Labs) integration.  It might be worth investigating.

Anyhow, thanks for the help and your thoughtful responses!

**Comment by naugtur** [comment]
I remember eliminating all mutations to options at some point. Maybe it's a regression, or I overlooked that.

Thanks for the tip. Feel free to come back whenever you need.
I'm looking forward to your comments on the split to modern and legacy browser support, once I get the initial version done.
